### PR TITLE
chore: update to half Sapphire

### PR DESCRIPTION
**Pull request by kyranet**
| ⚠️ This Pull Request needs https://github.com/skyra-project/klasa/pull/12 to be approved and merged beforehand. Once the linked Pull Request is merged, this one will receive its last commit to re-target to the fork's `settings` branch. |
| --- |

| ✔️ ~~This Pull Request needs https://github.com/sapphire-project/utilities/pull/61 to be approved, merged, and released beforehand. Once the linked Pull Request is merged, this one will receive its last commit to re-target to the latest version in npm.~~ |
| --- |

### CODE REVIEW

File: src/commands/Games/hungergames.ts [hunk 1]
```diff
@@ -179,7 +178,7 @@ export default class extends SkyraCommand {
 		const proceed = t(LanguageKeys.Commands.Games.HungerGamesResultProceed);
 		const panels = chunk(results, 5);
 
-		const texts = panels.map((panel) => `__**${header}:**__\n\n${panel.map((text) => `- ${text}`).join('\n')}\n\n_${proceed}_`);
+		const texts = panels.map((panel) => `__**${header}:**__\n\n${panel.map((text) => `- ${text}`).join('\n')}\n\n_${proceed}_`) as string[];
```
Comment (vladfrangu) [inline src/commands/Games/hungergames.ts:181]: Is this cast needed?

File: src/commands/General/Chat Bot Info/help.ts [hunk 2]
```diff
@@ -41,10 +41,10 @@ function sortCommandsAlphabetically(_: Command[], __: Command[], firstCategory:
 	flagSupport: true
 })
 export default class extends SkyraCommand {
-	public async init() {
+	public async onLoad() {
 		this.createCustomResolver('command', (arg, possible, message) => {
```
Comment (vladfrangu) [inline src/commands/General/Chat Bot Info/help.ts]: decorators?

File: src/commands/Pokemon/pokedex.ts [hunk 3]
```diff
@@ -112,7 +112,7 @@ export default class extends RichDisplayCommand {
 	 */
 	private getEvoChain(pokeDetails: DexDetails): string {
 		// Set evochain if there are no evolutions
-		let evoChain = `**${toTitleCase(pokeDetails.species)} ${pokeDetails.evolutionLevel ? `(${pokeDetails.evolutionLevel})` : ''}**`;
+		let evoChain = `**${toTitleCase(pokeDetails.species)} ${pokeDetails.evolutionLevel ? `(${pokeDetails.evolutionLevel})` : ''}**` as string;
```
Comment (vladfrangu) [inline src/commands/Pokemon/pokedex.ts:114]: ?

File: src/commands/Pokemon/type.ts [hunk 4]
```diff
@@ -82,15 +83,15 @@ export default class extends RichDisplayCommand {
 
 	private parseEffectiveMatchup(doubleEffectiveTypes: TypeEntry['doubleEffectiveTypes'], effectiveTypes: TypeEntry['effectiveTypes']) {
 		return doubleEffectiveTypes
-			.map((type) => `${type} (x4)`)
+			.map((type): string => `${type} (x4)`)
```
Comment (vladfrangu) [inline src/commands/Pokemon/type.ts:85]: deez casts?

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: src/commands/Games/hungergames.ts (hunk 1 above)
Comment (kyranet) [reply src/commands/Games/hungergames.ts:181]: Yes, because TypeScript 4.2.0 considers `texts` an array of templated types, on similar code:

image

File: src/commands/General/Chat Bot Info/help.ts (hunk 2 above)
Comment (kyranet) [reply src/commands/General/Chat Bot Info/help.ts]: The `category` resolver requires access to the `_fetchCommands` method from that class.

File: src/commands/Games/hungergames.ts (hunk 1 above)
Comment (vladfrangu) [reply src/commands/Games/hungergames.ts:181]: Ohhhh, I see

File: src/arguments/integer.ts [hunk 5]
```diff
@@ -13,7 +13,6 @@ export default class extends Argument {
 
 		const { min, max } = possible;
 
-		// @ts-expect-error 2341
```
Comment (favna) [inline src/arguments/integer.ts:16]: Yeetus deletus begones

File: src/commands/Pokemon/type.ts (hunk 4 above)
Comment (favna) [reply src/commands/Pokemon/type.ts:85]: 🌰🌰

File: src/events/reactions/rawReactionAddNotify.ts [hunk 6]
```diff
@@ -96,17 +96,17 @@ export default class extends Event {
 	}
 
 	private async fetchCount(data: LLRCData, emoji: string, id: string) {
-		const users = (await api(this.client).channels(data.channel.id).messages(data.messageID).reactions(emoji).get()) as APIUser[];
+		const users = (await api(this.context.client).channels(data.channel.id).messages(data.messageID).reactions(emoji).get()) as APIUser[];
```
Comment (favna) [inline src/events/reactions/rawReactionAddNotify.ts]: Can't we make this `api().channels` and add a DI based default client where we have the api stuff?

File: src/lib/SkyraClient.ts [hunk 7]
```diff
@@ -1,38 +1,34 @@
 /* eslint-disable @typescript-eslint/no-invalid-this */
 // Import all dependencies
```
Comment (favna) [inline src/lib/SkyraClient.ts]: Either split the imports in categories again, or remove all these comments if we're not splitting the imports anymore.

Personally I'm more in favour of the latter option.

File: src/lib/database/entities/GiveawayEntity.ts [hunk 8]
```diff
@@ -187,8 +187,8 @@ export class GiveawayEntity extends BaseEntity {
 		const t = await guild.fetchT();
 		if (state === States.Finished) {
 			this.#winners = await this.pickWinners();
+			this.#finished = true;
```
Comment (favna) [inline src/lib/database/entities/GiveawayEntity.ts:180]: Will the fact that it gets shifted up a line matter?

File: src/lib/structures/commands/ChannelConfigurationCommand.ts [hunk 9]
```diff
@@ -4,7 +4,7 @@ import type { CustomFunctionGet, GuildMessage, KeyOfType } from '#lib/types';
 import { PermissionLevels } from '#lib/types/Enums';
 import type { Nullish } from '@sapphire/utilities';
 import type { TextChannel } from 'discord.js';
-import type { CommandStore } from 'klasa';
+import { PieceContext } from 'klasa';
```
Comment (favna) [inline src/lib/structures/commands/ChannelConfigurationCommand.ts]:
```suggestion
import type { PieceContext } from 'klasa';
```

File: src/lib/structures/commands/ModerationCommand.ts [hunk 10]
```diff
@@ -8,7 +8,7 @@ import { cast, floatPromise } from '#utils/util';
 import { isNullOrUndefined } from '@sapphire/utilities';
 import type { User } from 'discord.js';
 import type { TFunction } from 'i18next';
-import type { CommandStore } from 'klasa';
+import { PieceContext } from 'klasa';
```
Comment (favna) [inline src/lib/structures/commands/ModerationCommand.ts]:
```suggestion
import type { PieceContext } from 'klasa';
```

File: src/lib/structures/commands/MusicCommand.ts [hunk 11]
```diff
@@ -1,5 +1,5 @@
 import { ENABLE_LAVALINK } from '#root/config';
-import type { CommandStore } from 'klasa';
+import { PieceContext } from 'klasa';
```
Comment (favna) [inline src/lib/structures/commands/MusicCommand.ts]:
```suggestion
import type { PieceContext } from 'klasa';
```

File: src/lib/structures/commands/RichDisplayCommand.ts [hunk 12]
```diff
@@ -1,6 +1,6 @@
 import type { GuildMessage } from '#lib/types';
 import type { PermissionResolvable } from 'discord.js';
-import type { CommandStore } from 'klasa';
+import { PieceContext } from 'klasa';
```
Comment (favna) [inline src/lib/structures/commands/RichDisplayCommand.ts]:
```suggestion
import type { PieceContext } from 'klasa';
```

File: src/lib/structures/commands/SelfModerationCommand.ts [hunk 13]
```diff
@@ -5,7 +5,7 @@ import { PermissionLevels } from '#lib/types/Enums';
 import type { KeyOfType } from '#lib/types/Utils';
 import { codeBlock } from '@sapphire/utilities';
 import type { TFunction } from 'i18next';
-import { CommandStore, Duration } from 'klasa';
+import { Duration, PieceContext } from 'klasa';
```
Comment (favna) [inline src/lib/structures/commands/SelfModerationCommand.ts]:
```suggestion
import type { PieceContext } from 'klasa';
import { Duration } from '@sapphire/time-utilities';
```

File: src/lib/types/Enums.d.ts [hunk 14]
```diff
@@ -102,6 +102,7 @@ export const enum PermissionLevels {
 	Moderator = 5,
 	Administrator = 6,
 	ServerOwner = 7,
+	PreBotOwner = 9,
```
Comment (favna) [inline src/lib/types/Enums.d.ts]: What's a PreBotOwner 🤔

**Review by favna — CHANGES_REQUESTED** [review]

File: src/lib/database/entities/GiveawayEntity.ts (hunk 8 above)
Comment (kyranet) [reply src/lib/database/entities/GiveawayEntity.ts:180]: No, it's more preference to do 2 sets together.

**Review by favna — DISMISSED** [review]

**Review by favna — DISMISSED** [review]

File: src/commands/General/Chat Bot Info/invite.ts [hunk 15]
```diff
@@ -25,10 +24,6 @@ export default class extends SkyraCommand {
 		return message.send(this.getEmbed(t, { permissions: true }));
 	}
 
-	public async init() {
-		if (this.client.application && !this.client.application.botPublic) this.permissionLevel = PermissionLevels.BotOwner;
```
Comment (vladfrangu) [inline src/commands/General/Chat Bot Info/invite.ts:29]: Why WAS this removed?

File: src/commands/Social/remindme.ts [hunk 16]
```diff
@@ -118,13 +118,16 @@ export default class extends SkyraCommand {
...
+		const tasks = this.context.client.schedules.queue.filter((task) => task.data && task.data.user === message.author.id);
 		if (!tasks.length) return message.sendTranslated(LanguageKeys.Commands.Social.RemindMeListEmpty);
 
 		const display = new UserRichDisplay(
 			new MessageEmbed()
 				.setColor(await DbSet.fetchColor(message))
-				.setAuthor(this.client.user!.username, this.client.user!.displayAvatarURL({ size: 128, format: 'png', dynamic: true }))
+				.setAuthor(
+					this.context.client.user!.username,
+					this.context.client.user!.displayAvatarURL({ size: 128, format: 'png', dynamic: true })
```
Comment (vladfrangu) [inline src/commands/Social/remindme.ts]: extract client from context and user from client

**Review by vladfrangu — CHANGES_REQUESTED** [review]

File: src/commands/General/Chat Bot Info/invite.ts (hunk 15 above)
Comment (kyranet) [reply src/commands/General/Chat Bot Info/invite.ts:29]: Because it always posts the production bot's invite link.

**Review by vladfrangu — APPROVED** [review]

File: src/commands/Google/gimage.ts [hunk 17]
```diff
@@ -34,7 +34,7 @@ export default class extends RichDisplayCommand {
 			queryGoogleCustomSearchAPI<CustomSearchType.Image>(message, CustomSearchType.Image, query)
 		]);
 
-		if (!items || !items.length) throw t(handleNotOK(GoogleResponseCodes.ZeroResults, message.client));
+		if (!items || !items.length) throw t(handleNotOK(GoogleResponseCodes.ZeroResults));
```
Comment (vladfrangu) [inline src/commands/Google/gimage.ts:37]: Use ?.

File: src/commands/Google/gsearch.ts [hunk 18]
```diff
@@ -33,7 +33,7 @@ export default class extends RichDisplayCommand {
 			queryGoogleCustomSearchAPI<CustomSearchType.Search>(message, CustomSearchType.Search, query)
 		]);
 
-		if (!items || !items.length) throw t(handleNotOK(GoogleResponseCodes.ZeroResults, message.client));
+		if (!items || !items.length) throw t(handleNotOK(GoogleResponseCodes.ZeroResults));
```
Comment (vladfrangu) [inline src/commands/Google/gsearch.ts:36]: bop
